### ISSUE TITLE
Blank lines in TS code cause breakpoints to move upon browser refresh.

### ISSUE THREAD

**Issue by weinand** [Microsoft/vscode-chrome-debug#296]
_From @jeffsim on November 25, 2016 15:34_

- VSCode Version: 1.7.2
- OS Version: Windows 10
- Debugger for Chrome extension version: 2.2.2

I drop a breakpoint into typescript code which will get hit after the page loads as expected.  I continue running in VS Code and then hit 'f5' in the browser; the breakpoint is hit again in VS Code, but the red breakpoint dot in VS Code moves up to an earlier line in the code.  The distance it moves up is equal to the number of empty lines before the breakpoint's original position.

Here's what it looks like on first run:

![On first run](https://www.dropbox.com/s/lvj0r3uq5gcw5t4/firstRun.png?raw=1)

And here's what it looks like after continuing and then hitting refresh in Chrome:

![On first run](https://www.dropbox.com/s/dap0vm0l72oxkui/afterRefresh.png?raw=1)

Apologies if this is caused by an obvious error on my part; this seems like something everyone would have hit, so I've created a minimum repro case of the issue to help demonstrate it.  I'd greatly appreciate any assistance in understanding what's going on.

Steps to Reproduce:

1. Download the ~4kb zipped project from here: https://www.dropbox.com/s/e8xu4157n9j3hxa/tsNSRepro.zip?dl=0
2. npm install to get the dependencies, and ensure Chrome and the Debugger for Chrome extension are both installed
3. Place a breakpoint on the "console.log" line in testApp.ts.
4. Build the project and run.  The breakpoint should be hit and appear as per first picture above
5. F5 to continue running in VS Code
6. In the browser, hit F5 to refresh the page
7. Note that in vs code the breakpoint is again hit, but the red dot has moved up the page.

Try adding/removing blank lines and repeat the above, and note how the distance the red dot moves changes accordingly.

Here's the typescript file in question:
```
// NOTE: Blank lines screw up breakpoint loc on F5.


setTimeout(function() {
    console.log("i am a test");
}, 1000);
```

... and here's the gulpfile
```
var gulp = require("gulp"),
    tsc = require("gulp-typescript"),
    sourcemaps = require("gulp-sourcemaps"),
    del = require("del"),
    runSequence = require("run-sequence");

gulp.task("clean", function () {
    return del([ "testApp.js.map", "testApp.js" ]);
});

gulp.task("build-testApp", function (cb) {
    var ts = tsc.createProject("tsconfig.json");
    return gulp.src(["testApp.ts"])
        .pipe(sourcemaps.init())
        .pipe(ts())
        .pipe(sourcemaps.write(".", { includeContent: false, sourceRoot: "." })) 
        .pipe(gulp.dest("."))
});

gulp.task("build-all", function (cb) {
    runSequence("clean", "build-testApp", cb);
});
```

And finally, here's the diagnostic logging output after performing the above steps and F5'ing a couple of times - nothing seems to imply a failure to map breakpoints...

```
Verbose logs are written to C:\Users\jeffs\.vscode\extensions\msjsdiag.debugger-for-chrome-2.2.2\vscode-chrome-debug.txt
8:04:15 AM, 11/25/2016
OS: win32 ia32
Adapter node: v6.5.0 ia32
vscode-chrome-debug-core: 3.6.3
debugger-for-chrome: 2.2.1
spawn('C:\Program Files (x86)\Google\Chrome\Application\chrome.exe', ["--remote-debugging-port=9222","--no-first-run","--no-default-browser-check","--disable-session-crashed-bubble","--disable-infobars","http://127.0.0.1:1025/index.html"])
Discovering targets via http://127.0.0.1:9222/json
Target userAgent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.99 Safari/537.36
Paths.scriptParsed: resolved http://127.0.0.1:1025/testApp.js to c:\Users\jeffs\Desktop\tsRepro\testApp.js. webRoot: c:\Users\jeffs\Desktop\tsRepro
SourceMaps.getMapForGeneratedPath: Finding SourceMap for c:\Users\jeffs\Desktop\tsRepro\testApp.js by URI: testApp.js.map and webRoot: c:\Users\jeffs\Desktop\tsRepro
SourceMaps.loadSourceMapContents: Reading local sourcemap file from c:\Users\jeffs\Desktop\tsRepro\testApp.js.map
SourceMap: creating for c:\Users\jeffs\Desktop\tsRepro\testApp.js
SourceMap: sourceRoot: .
SourceMap: sources: ["testApp.ts"]
SourceMap: webRoot: c:\Users\jeffs\Desktop\tsRepro
SourceMap: resolved sourceRoot . -> c:\Users\jeffs\Desktop\tsRepro
SourceMaps.scriptParsed: c:\Users\jeffs\Desktop\tsRepro\testApp.js was just loaded and has mapped sources: ["c:\\Users\\jeffs\\Desktop\\tsRepro\\testApp.ts"]
SourceMaps.setBP: Mapped c:\Users\jeffs\Desktop\tsRepro\testApp.ts to c:\Users\jeffs\Desktop\tsRepro\testApp.js
SourceMaps.setBP: Mapped c:\Users\jeffs\Desktop\tsRepro\testApp.ts:9:1 to c:\Users\jeffs\Desktop\tsRepro\testApp.js:3:5
Paths.setBP: Resolved c:\Users\jeffs\Desktop\tsRepro\testApp.js to http://127.0.0.1:1025/testApp.js
SourceMaps.setBP: Mapped c:\Users\jeffs\Desktop\tsRepro\testApp.js:3:5 to c:\Users\jeffs\Desktop\tsRepro\testApp.ts:9
i am a test
Paths.scriptParsed: resolved http://127.0.0.1:1025/testApp.js to c:\Users\jeffs\Desktop\tsRepro\testApp.js. webRoot: c:\Users\jeffs\Desktop\tsRepro
SourceMaps.scriptParsed: c:\Users\jeffs\Desktop\tsRepro\testApp.js was just loaded and has mapped sources: ["c:\\Users\\jeffs\\Desktop\\tsRepro\\testApp.ts"]
i am a test
Paths.scriptParsed: resolved http://127.0.0.1:1025/testApp.js to c:\Users\jeffs\Desktop\tsRepro\testApp.js. webRoot: c:\Users\jeffs\Desktop\tsRepro
SourceMaps.scriptParsed: c:\Users\jeffs\Desktop\tsRepro\testApp.js was just loaded and has mapped sources: ["c:\\Users\\jeffs\\Desktop\\tsRepro\\testApp.ts"]
i am a test
```
Other supporting files (tsconfig.json etc) can be found in the above linked zip file.

Thanks,
Jeff

_Copied from original issue: Microsoft/vscode#16066_

**Comment by weinand** [comment]
_From @isidorn on November 25, 2016 16:52_

The breakpoint updates are coming from the debug adapter moving to @weinand

**Comment by auchenberg** [comment]
Sounds like the previous sourcemap was used/cached, and the newly generate map needs to fetched?

**Comment by jeffsim** [comment]
That's a reasonable guess @auchenberg, but I believe it repros even if you remove any retranspilation/precaching scenarios.  e.g. the following repro steps cause it to still happen:

1. Completely reset chrome (control+shift+delete from within it.  NOTE: this'll _reset_ chrome, so don't do this unless you're okay losing history etc! :)).
2. Run the test steps 4-7; it repros.  Note that step 4 cleans out any old sourcemaps and JS
3. stop debugging, and re-run steps 5-7.  Still repros.

The above continues to be the case even if you re-run multiple times, try closing out chrome, try closing out vs code, and try re-resetting chrome and not re-building the test.

It's certainly possible that the sourcemap is still getting cached at some other point that I'm unaware of, but I'd naively assume that the above steps would catch all points.

Thanks again,
Jeff

**Comment by jeffsim** [comment]
In case it sheds any further light, here's the part of the verbose diagnostic log starting from immediately after step 5 above;

```
  ›To target: "{\"id\":12,\"method\":\"Debugger.resume\"}"
  ›From target: {"id":12,"result":{}}
  ›To client: {"seq":0,"type":"response","request_seq":10,"command":"continue","success":true}
  ›From target: {"method":"Debugger.resumed","params":{}}
  ›From target: {"method":"Console.messageAdded","params":{"message":{"source":"console-api","level":"log","text":"i am a test","url":"http://127.0.0.1:1025/testApp.js","line":3,"column":13}}}
  ›From target: {"method":"Runtime.consoleAPICalled","params":{"type":"log","args":[{"type":"string","value":"i am a test"}],"executionContextId":2,"timestamp":1480175786764.802979,"stackTrace":{"callFrames":[{"functionName":"","scriptId":"36","url":"http://127.0.0.1:1025/testApp.js","lineNumber":2,"columnNumber":12}]}}}
i am a test
  ›To target: "{\"id\":13,\"method\":\"Page.configureOverlay\",\"params\":{}}"
  ›From target: {"id":13,"result":{}}
  ›From target: {"method":"Runtime.executionContextDestroyed","params":{"executionContextId":2}}
  ›From target: {"method":"Runtime.executionContextsCleared","params":{}}
  ›From target: {"method":"Runtime.executionContextCreated","params":{"context":{"id":4,"origin":"http://127.0.0.1:1025","name":"","auxData":{"isDefault":true,"frameId":"22596.1"}}}}
  ›From target: {"method":"Debugger.scriptParsed","params":{"scriptId":"42","url":"http://127.0.0.1:1025/testApp.js","startLine":0,"startColumn":0,"endLine":7,"endColumn":0,"executionContextId":4,"hash":"FE4592A39B02DBE82D0345F426E3F5213662CB25","executionContextAuxData":{"isDefault":true,"frameId":"22596.1"},"isLiveEdit":false,"sourceMapURL":"testApp.js.map","hasSourceURL":false}}
Paths.scriptParsed: resolved http://127.0.0.1:1025/testApp.js to c:\Users\jeffs\Desktop\tsRepro\testApp.js. webRoot: c:\Users\jeffs\Desktop\tsRepro
SourceMaps.scriptParsed: c:\Users\jeffs\Desktop\tsRepro\testApp.js was just loaded and has mapped sources: ["c:\\Users\\jeffs\\Desktop\\tsRepro\\testApp.ts"]
  ›From target: {"method":"Debugger.breakpointResolved","params":{"breakpointId":"/http:\\/\\/127\\.0\\.0\\.1:1025\\/testApp\\.js/:2:4","location":{"scriptId":"42","lineNumber":2,"columnNumber":4}}}
  ›To client: {"seq":0,"type":"event","event":"breakpoint","body":{"reason":"new","breakpoint":{"id":1000,"verified":true,"line":3,"column":5}}}
  ›From target: {"method":"Debugger.paused","params":{"callFrames":[{"callFrameId":"{\"ordinal\":0,\"injectedScriptId\":4}","functionName":"","functionLocation":{"scriptId":"42","lineNumber":1,"columnNumber":20},"location":{"scriptId":"42","lineNumber":2,"columnNumber":4},"scopeChain":[{"type":"local","object":{"type":"object","className":"Object","description":"Object","objectId":"{\"injectedScriptId\":4,\"id\":1}"},"startLocation":{"scriptId":"42","lineNumber":1,"columnNumber":20},"endLocation":{"scriptId":"42","lineNumber":3,"columnNumber":1}},{"type":"global","object":{"type":"object","className":"Window","description":"Window","objectId":"{\"injectedScriptId\":4,\"id\":2}"}}],"this":{"type":"object","className":"Window","description":"Window","objectId":"{\"injectedScriptId\":4,\"id\":3}"}}],"reason":"other","hitBreakpoints":["/http:\\/\\/127\\.0\\.0\\.1:1025\\/testApp\\.js/:2:4"]}}
  ›To target: "{\"id\":14,\"method\":\"Page.configureOverlay\",\"params\":{\"message\":\"Paused in Visual Studio Code\"}}"
  ›To client: {"seq":0,"type":"event","event":"stopped","body":{"reason":"breakpoint","threadId":1}}
  ›From target: {"id":14,"result":{}}
  ›From client: threads(undefined)
  ›To client: {"seq":0,"type":"response","request_seq":11,"command":"threads","success":true,"body":{"threads":[{"id":1,"name":"Thread 1"}]}}
  ›From client: stackTrace({"threadId":1,"startFrame":0,"levels":20})
  ›To client: {"seq":0,"type":"response","request_seq":12,"command":"stackTrace","success":true,"body":{"stackFrames":[{"id":1000,"name":"(anonymous function)","source":{"name":"testApp.ts","path":"c:\\Users\\jeffs\\Desktop\\tsRepro\\testApp.ts","sourceReference":0}**,"line":9,"column":5**}]}}
  ›From client: scopes({"frameId":1000})
  ›To client: {"seq":0,"type":"response","request_seq":13,"command":"scopes","success":true,"body":{"scopes":[{"name":"Local","variablesReference":1000,"expensive":false,"column":20,"line":1,"endColumn":1,"endLine":3},{"name":"Global","variablesReference":1001,"expensive":true}]}}
  ›From client: variables({"variablesReference":1000})
  ›To target: "{\"id\":15,\"method\":\"Runtime.getProperties\",\"params\":{\"objectId\":\"{\\\"injectedScriptId\\\":4,\\\"id\\\":1}\",\"ownProperties\":false,\"accessorPropertiesOnly\":true,\"generatePreview\":true}}"
  ›To target: "{\"id\":16,\"method\":\"Runtime.getProperties\",\"params\":{\"objectId\":\"{\\\"injectedScriptId\\\":4,\\\"id\\\":1}\",\"ownProperties\":true,\"accessorPropertiesOnly\":false,\"generatePreview\":true}}"
  ›From target: {"id":15,"result":{"result":[]}}
  ›From target: {"id":16,"result":{"result":[]}}
  ›To client: {"seq":0,"type":"response","request_seq":14,"command":"variables","success":true,"body":{"variables":[{"name":"this","value":"Window","type":"Window","variablesReference":1002,"evaluateName":"this"}]}}
```

Without knowing too much about what all that means, these are interesting lines that stand out to me:

```
SourceMaps.scriptParsed: c:\Users\jeffs\Desktop\tsRepro\testApp.js was just loaded and has mapped sources: ["c:\\Users\\jeffs\\Desktop\\tsRepro\\testApp.ts"]
  ›From target: {"method":"Debugger.breakpointResolved","params":{"breakpointId":"/http:\\/\\/127\\.0\\.0\\.1:1025\\/testApp\\.js/:2:4","location":{"scriptId":"42","lineNumber":2,"columnNumber":4}}}
  ›To client: {"seq":0,"type":"event","event":"breakpoint","body":{"reason":"new","breakpoint":{"id":1000,"verified":true,"line":3,"column":5}}}
  ›From target: {"method":"Debugger.paused","params":{"callFrames":[{"callFrameId":"{\"ordinal\":0,\"injectedScriptId\":4}","functionName":"","functionLocation":{"scriptId":"42","lineNumber":1,"columnNumber":20},"location":{"scriptId":"42","lineNumber":2,"columnNumber":4},"scopeChain":[{"type":"local","object":{"type":"object","className":"Object","description":"Object","objectId":"{\"injectedScriptId\":4,\"id\":1}"},"startLocation":{"scriptId":"42","lineNumber":1,"columnNumber":20},"endLocation":{"scriptId":"42","lineNumber":3,"columnNumber":1}},{"type":"global","object":{"type":"object","className":"Window","description":"Window","objectId":"{\"injectedScriptId\":4,\"id\":2}"}}],"this":{"type":"object","className":"Window","description":"Window","objectId":"{\"injectedScriptId\":4,\"id\":3}"}}],"reason":"other","hitBreakpoints":["/http:\\/\\/127\\.0\\.0\\.1:1025\\/testApp\\.js/:2:4"]}}
```

in particular,the first "To Client:" line, which implies a _new_ breakpoint has been set (?).

Cheers,
Jeff

**Comment by jeffsim** [comment]
Different repro steps that show the red breakpoint dot dancing back and forth between lines 3 and 9 with the same sourcemap:

1. Build app. drop a breakpoint on the 'console.log' line
2. F5 in vs code.  Wait until it hits the breakpoint.
3. Hit F5 in the _browser_ to refresh the page
4. Note that the breakpoint moves by itself to line 3
5. Stop debugging in vs code; breakpoint remains on wrong line
6. Hit F5 in vscode; notice that the breakpoint moves _back_ to line 9, and it gets hit correctly.
7. Jump back to step 3, and repeat ad infinitum.

I'm pretty sure that's not a precached sourcemaps scenario.  Weird!

Jeff

**Comment by roblourens** [comment]
Thanks for the detailed report, I think this is fixed in 2.3.0. Try it out and let me know!

**Comment by jeffsim** [comment]
Works like a charm now!  Thanks for the amazingly quick turnaround!
